---
title: "Results - B.1.1.7,  “Alpha variant”"
author: ""
date: "21/06/2021"
output: 
  github_document
#pdf_document 
---
  
  ```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning =F)
```
Phylogeographic mapping

The analyses take a dated phylogeny exported from NextStrain as a starting point for quantifying imports and local transmission of Sars-CoV-2 B.1.1.7 in Norway. 
Each isolate is assigned to one of two binary categories: "Norway" or "Rest of the world (RoW)". The output is presented in Fig.1 and ancestral character estimation performed on the dated phylogeny. 

```{r remedy001, include=FALSE}
### . . . . . . . . . . . . . . .Notes . . . . . . . . . . . . . . . . . . . . .
### . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ..
### Sections that takes time to run are commented  out                      ####
### When a new build is available these should be un-commented, the         ####
### result saved, and re-commented. ylim's may have to be adjusted          ####

#   ____________________________________________________________________________
#   Loading required packages                                               ####

library(ggplot2)
library(phytools)
library(ape)
library(castor)
library(ips)
library(data.table)
library(lubridate)
library(ggtree)
library(treedater)
library(viridis)
library(grid)
library(BactDating)
library(pbapply)
library(LineageHomology)
library(hexbin)

#   ____________________________________________________________________________
#   Setup paths - replace these to run locally                              ####

#The folder names are old, but they contain the newest data.
path_Main = "~/Dropbox/Covid/UK_03_24/"
path_Results = "~/Dropbox/Covid/UK_03_24/Results/"
path_Rscripts = "~/Dropbox/Covid/Rfunctions/"
path_Data =  "~/Dropbox/Covid/UK_03_24/Data/"


#   ____________________________________________________________________________
#   Read required data                                                      ####
# Some newer builds including sequences not uploaded to Gisaid lack 'Norway' in sample ID.
# Fixed this in both .tsv and .nwk files using the script:
# ~/Dropbox/Covid/UK_03_24/replace_words.py and the conversion file:
# ~/Dropbox/Covid/UK_03_24/conversion_HF_isolnames.csv

metadat = fread(paste0(path_Main,"nextstrain_groups_niph_ncov_UK-centric-2021-06-21_metadata_NN.tsv"))
metadata = metadat
tree = read.tree(paste0(path_Main,"nextstrain_groups_niph_ncov_UK-centric-2021-06-21_timetree_NN.nwk"))

#   ____________________________________________________________________________
#   Analysis begins here                                                    ####


#Extract the most recent branches
plot.phylo(tree,show.tip.label = F,type="cladogram")
nodelabels(cex=0.5)
tree = extract.clade(tree, node=5846)

#Expand zero length branches
set.seed(100)
#tree = multi2di(tree)
tree$edge.length[tree$edge.length<=0]=min(tree$edge.length[tree$edge.length>0])
tree=ladderize(tree)#reorder.phylo(tree)
plot.phylo(tree,show.tip.label = F,edge.width=2)


Locations = grepl("Norway",tree$tip.label)
Locations[Locations=="TRUE"]="Norway"
Locations[Locations=="FALSE"]="RoW"
table(Locations)
names(Locations)=tree$tip.label


## Setup transition matrix estimated using ML and the assumption 10X rate to Norway. DO RUN WHENEVER NEW BUILD ADDED
#LineageHomology::FitTransition10x(tree,Locations,model=c(rbind(c(0,1,1,0)))) #The 10X rate is added internally
## Update below matrix with new estimates from Lineage Homology
transition_matrix = rbind(c(-0.1159325, 0.1159325),c(1.1593248, -1.1593248))
## Old Estimate from 2021-06-15: transition_matrix = rbind(c(-0.124796, 0.124796),c(1.247960, -1.247960))
## Old Estimate from 2021-06-07: transition_matrix = rbind(c(-0.1226109, 0.12261091.2261085),c(1.22610851.2261085, -1.2261085)) #Write up from the print above
## Old Estimate from 2021-05-31: transition_matrix = rbind(c(-0.1202222, 0.1202222),c(1.2022219, -1.2022219)) #Write up from the print above
## Old Estimate from 2021-05-25: transition_matrix = rbind(c(-0.1370393, 0.1370393),c(1.3703926, -1.3703926)) #Write up from the print above
## Old Estimate from 2021-05-18: transition_matrix = rbind(c(-0.1436812, 0.1436812),c(1.4368116, -1.4368116)) #Write up from the print above
## Old Estimate from 2021-05-10 transition_matrix = rbind(c(-0.1388084, 0.1388084),c(1.3880840, -1.3880840)) #Write up from the print above
## Old Estimate from 2021-05-03 transition_matrix = rbind(c(-0.1319801, 0.1319801),c(1.3198013, -1.3198013)) #Write up from the print above
## Old Estimate from 2021-04-29: transition_matrix = rbind(c(-0.1353298, 0.1353298),c(1.3532984, -1.3532984)) #Write up from the print above
## Old Estimate from 2021-04-27: transition_matrix = rbind(c(-0.07326712, 0.07326712),c(0.73267120, -0.73267120)) #Write up from the print above
## Old Estimate from 2021-04-20: transition_matrix = rbind(c(-0.07228033 ,0.07228033),c(0.72280328,-0.72280328))
fit_given = fitMk(tree, Locations,fixedQ = transition_matrix, pi=c(0,1)); AIC(fit_given)
transition_matrix = summary(fit_given)$Q
#Setup locations for castor
locations_for_castor = as.numeric(as.factor(Locations))
names(locations_for_castor)=names(Locations)

#   ____________________________________________________________________________
#   Ancestral state reconstruction                                          ####

#Reconstruct ancestral states using maximum likelihood with the fixed transition matrix above.

castor = castor::asr_mk_model(tree,tip_states =locations_for_castor,
                              transition_matrix=transition_matrix,include_ancestral_likelihoods = T)


#Extract the most recent dates.
mrsd = max(metadat$`Collection Data`)
dec_date = decimal_date(max(metadat$`Collection Data`))
start_date = dec_date-max(tipHeights(tree))
most_recent_date = ymd(mrsd)
most_recent_date = decimal_date(most_recent_date)
start_time = most_recent_date-max(nodeHeights(tree))
time_end = most_recent_date-start_time
time_end = start_time+max(nodeHeights(tree))

#Find the labels of the taxa that are observed in Norway. 
Norwegian_tips = names(Locations[Locations=="Norway"])


### . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ..
### Sections that takes time to run are commented  out, like below          ####
### New results should be saved and loaded once new builds are produced     ####

#Lineage description from the maximum likelihood mappings obtained by sampling from node probabilities.
## Run below command for new builds. If results needed fast use 10 replicates below, and rerun with e.g. 100.
#multi_counts = pbreplicate(10, LineageHomology_w_uncertainty(tree, ace_nodes = castor$ancestral_likelihoods, give_tips = Norwegian_tips,ace_tips = to.matrix(Locations,seq=c("Norway","RoW")),start_time = start_date))
## Run below command for new builds
#save(multi_counts, file=paste0(path_Data,"count_ML_UK_100.Rdata"))
load(paste0(path_Data,"count_ML_UK_100.Rdata"))

#Lineage descriptions using the state with more than 50% probability for all nodes.
## Run below commands for new builds
#Result = LineageHomology(tree, ace_nodes = castor$ancestral_likelihoods, give_tips = Norwegian_tips,ace_tips = to.matrix(Locations,seq=c("Norway","RoW")),start_time = start_date)
#save(Result,file=paste0(path_Data,"count_once_UK.Rdata"))
load(paste0(path_Data,"count_once_UK.Rdata"))
```

```{r, echo=FALSE, fig.width=13, fig.height=20, fig.cap="Test"}
cols<-setNames( c("#B3E2CD","#FDCDAC"),unique(Locations))
plot.phylo(tree,cex=0.3, mar=c(5.1,0.2,0.2,0.2),show.tip.label = F,edge.width=2)
LineageHomology::axisPhyloDate(1,root.time= decimal_date(mrsd)-max(tipHeights(tree)), backward = F, lwd=4,cex=5)
nodelabels(pie=castor$ancestral_likelihoods,cex=0.2,piecol=cols)
tiplabels(pie=to.matrix(Locations,seq=c("Norway","RoW")),cex=0.2,piecol=cols)
legend("bottomleft",c("Norway", "RoW"), col = c("#66C2A5","#FC8D62"), pch=20, cex = 5, pt.cex=5, bty="n")
```

##### Fig. 1. B.1.1.7 phylogeny colored by binary geographic mappings (Norway, rest of the world [RoW]). 

Quantifying introductions and local transmission

Below is a visualization of the size of each transmission lineage detected in Norway. 

```{r, echo=FALSE, fig.width=8, fig.height=8}
all_lineages = Result[[2]]
#Plot lineages:
library(treemap)
library(RColorBrewer)
library(ggtree)
#Treemap lineages.
atreemap = data.frame(group=paste0("G:",1:length(all_lineages)," S: ", all_lineages, " mrca: ",decimal2Date(Result$`MRCA's`)),Value=all_lineages)
treemap(atreemap,index=("group"),vSize="Value", type="index", palette="Paired")
```
##### Fig. 2. B.1.1.7 transmission lineage sizes in Norway. G denotes the number assigned to each lineage, S denotes the size of the lineage and TMRCA denotes the age of the lineage.


```{r,include=F}
##MRCA's lineages size.
tip_list = Norwegian_tips
dates = decimal_date(metadat$`Collection Data`)
names(dates)=metadata$Strain

#Import versus local transmission. 
## Run below command for new builds
#Result_with_uncertainty = Relative_load_import_multi(tree, multi_counts,dates, start_date, tip_list=Norwegian_tips)
#save(Result_with_uncertainty,file=paste0(path_Data,"Relative_load_100.Rdata")) 
load(paste0(path_Data,"Relative_load_100.Rdata"))
start_time = start_date
#Set windows for counting importations and local transmission in.
weeks = seq(start_time, time_end, by = 1/52)
week_time=weeks[1:(length(weeks)-1)]+1/(52*2) #This is only for returning results at midpoint of weeks.
weekly_importations = apply(Result_with_uncertainty$Import,2, FUN="mean")
weekly_local_tranmissions = apply(Result_with_uncertainty$LC,2, FUN="mean")
sum_cases = weekly_importations+weekly_local_tranmissions

#Plotting method
imports = t(apply(Result_with_uncertainty$Import, 2,quantile,c(0.025,0.5,0.975)))
exports = t(apply(Result_with_uncertainty$LC, 2,quantile,c(0.025,0.5,0.975)))
dat=data.frame(cbind(imports,exports),dates=decimal2Date(week_time))
colnames(dat)=c("i2.5","i50","i97.5","e2.5","e50","e97.5","dates")

#This uses the mean and discards uncertainty.
ggdata = data.frame(dates = as.Date(date_decimal(week_time)),
                    fraction=c(weekly_importations/sum_cases,weekly_local_tranmissions/sum_cases),
                    group=c(rep("Import",length(week_time)),
                            rep("Local transmission",length(week_time))))
```

##### Table 1. Estimated number of infections in Norway resulting from import and local transmission based on the phylogeographic analyses in Figure 1.

```{r, echo=F,fig.width=10, fig.height=10}
imports_temp = unlist(multi_counts[1,])[seq(1,length(multi_counts[1,])*2,by=2)]
local_temp = unlist(multi_counts[1,])[seq(2,length(multi_counts[1,])*2,by=2)]
relative_temp = imports_temp/(imports_temp+local_temp)
c_one = quantile(imports_temp,c(0.025,0.50,0.975))
c_two = quantile(local_temp,c(0.025,0.50,0.975))
c_three = quantile(relative_temp,c(0.025,0.50,0.975))
imp_loc = rbind(c_one,c_two, c_three)
rownames(imp_loc)=c("Import", "Local transmission", "Import / Total")
knitr::kable(imp_loc, booktabs = T, digits = 4)
```
Next, we estimate the fraction of introductions relative to local transmission over time. In these and subsequent analyses it is important to keep detection lag into account (see https://science.sciencemag.org/content/371/6530/708). There will always be a delay between the introduction and detection, which varies from case to case. In the figures below, the last four weeks prior to the time of sampling of the most recent included isolate is shaded to indicate that estimates in this period is highly uncertain.

```{r,echo=F, fig.width=7,fig.height=7}
g1 = ggplot(data=ggdata, aes(x=dates, y=fraction, fill=group)) + 
  geom_area(alpha=0.6 , size=1, colour="black")+
  theme_minimal(base_size=20)+scale_fill_brewer(palette="Set1")+
  scale_x_date(date_breaks="1 week")+
  theme(axis.text.x=element_text(angle=60, hjust=1))+
  ylab("")+xlab("")+
  annotate("rect",xmin=as.Date(decimal2Date(time_end-1/13)),xmax=as.Date(dat$dates[length(dat$dates)]),ymin=-Inf,ymax=Inf,alpha=0.2)+
  theme(legend.position = c(0.55,0.15), legend.title=element_blank())
g1
```

##### Fig. 3. Relative quantification of infection source (new introduction / import vs local transmission).

The same output is visualized as absolute numbers in Fig. 3 below. It should be noted that absolute estimates will be significantly shaped by sampling density.

```{r,include=F}

g1 = ggplot(dat, aes(x=dates,y=i50))+geom_line(col="Red")+theme_minimal(base_size=20)+
  theme(axis.title.x = element_blank(), axis.text.x = element_blank())+
  ylim(0,30)+
  geom_ribbon(aes(ymin=i2.5, ymax=i97.5,alpha=0.2), fill="lightblue")+
  scale_x_date(date_breaks="1 week",limits =c(as.Date("2020-11-30"), as.Date(dat$dates[length(dat$dates)])-7),expand=c(0,0))+
  ylab("Import")+
  theme(legend.position = "none")+annotate("rect",xmin=as.Date(decimal2Date(time_end-1/13)),xmax=as.Date(dat$dates[length(dat$dates)]-7),ymin=-Inf,ymax=Inf,alpha=0.2)
g1

g2 = ggplot(dat, aes(x=dates,y=e50))+geom_line(col="Darkgreen")+theme_minimal(base_size=20)+ylim(0,400)+
  scale_x_date(date_breaks="1 week",limits =c(as.Date("2020-11-30"),as.Date(dat$dates[length(dat$dates)])-7), expand=c(0,0))+
  geom_ribbon(aes(ymin=e2.5, ymax=e97.5,alpha=0.2), fill="lightblue")+
  theme(axis.text.x=element_text(angle=60, hjust=1))+scale_fill_continuous("Grey")+ylab("Local transmission")+xlab("")+
  theme(legend.position = "none")+annotate("rect",xmin=as.Date(decimal2Date(time_end-1/13)),xmax=as.Date(dat$dates[length(dat$dates)]-7),ymin=-Inf,ymax=Inf,alpha=0.2)
g2
```


```{r,echo=F,fig.width=10, fig.height=10}
grid.draw(rbind(ggplotGrob(g1), ggplotGrob(g2), size = "last"))
```

##### Fig. 4. Absolute quantification of imports and local transmission events. 

Perhaps the most interesting aspect of quantifying introductions, is to investigate the fate of the introduced viruses. Mandatory testing at the border and/or mandatory quaranteens etc will obviously have a bearing on the fate of new introductions from abroad. The plot below visualizes the number of infections in Norway resulting from each
identified virus import event over time, stratified by inferred TMRCA of each introduced lineage.

```{r,echo=F,fig.width=9, fig.height=9}
c2 = data.frame(decimal2Date(cbind(Result$`MRCA's`)),Result$Lineage_sizes)
colnames(c2)=c("MRCA","Lineagesize")
g1 = ggplot(c2, aes(MRCA, Lineagesize))+
  geom_hex()+
  theme_bw(base_size=20)+
  theme(legend.position=c(0.85,0.85))+
  scale_fill_viridis()+
  scale_x_date(date_breaks="1 week")+
  theme(axis.text.x=element_text(angle=60, hjust=1))+
  xlab("Estimated TMRCA")+ylab("Estimated linage size")+
  annotate("rect",xmin=as.Date(decimal2Date(time_end-1/13)),xmax=as.Date(dat$dates[length(dat$dates)]),ymin=-Inf,ymax=Inf,alpha=0.2)
g1

```

##### Fig. 5. Transmission lineage sizes as a function of time of introduction. “TMRCA” = time of most recent common ancestor.


```{r,include=F}
# Create tabular output of TMRCAs and lineage sizes

mrcas <- Result$`MRCA's`
mrcas <- decimal2Date(cbind(mrcas))
mrcas <- as.data.frame(mrcas)
Lin_sizes <- Result$Lineage_sizes
Lin_sizes <- as.data.frame(Lin_sizes)
mrca_sizes <- cbind(mrcas, Lin_sizes)
write.table(mrca_sizes, file=("~/Dropbox/Covid/UK_03_24/MRCA_lineageSize.txt"))
write.csv(mrca_sizes, file=("/home/vehuardo/Dropbox/Covid/Alpha_vs_Delta/alpha.csv"))

```
